File: src/Web/Dart/composite.dart
abstract interface class Node {
  int get size;
}

final class FileLeaf implements Node {
  FileLeaf(this.bytes);

  final int bytes;

  @override
  int get size => bytes;
}

final class FolderComposite implements Node {
  FolderComposite(this.children);

  final List<Node> children;

  @override
  int get size => children.fold(0, (total, child) => total + child.size);
}

void main() {
  final Node readme = FileLeaf(2);
  final Node docs = FolderComposite([FileLeaf(3), FileLeaf(5)]);
  final Node root = FolderComposite([readme, docs]);

  print('leaf=${readme.size}');
  print('docs=${docs.size}');
  print('root=${root.size}');
}

File: src/Web/JavaScriptJS/composite.js
class FileLeaf {
  constructor(bytes) {
    this.bytes = bytes;
  }

  size() {
    return this.bytes;
  }
}

class FolderComposite {
  constructor(children) {
    this.children = children;
  }

  size() {
    return this.children.reduce((total, child) => total + child.size(), 0);
  }
}

const readme = new FileLeaf(2);
const docs = new FolderComposite([new FileLeaf(3), new FileLeaf(5)]);
const root = new FolderComposite([readme, docs]);

console.log(`leaf=${readme.size()}`);
console.log(`docs=${docs.size()}`);
console.log(`root=${root.size()}`);
